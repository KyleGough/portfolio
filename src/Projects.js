import React, { useEffect, useState } from 'react';
import Section from './components/Section';
import ProjectItem from './components/ProjectItem';

export default function Projects() {
    const [filter, setFilter] = useState('All');

    useEffect(() => {
        document.title = 'Project List - Kyle Gough';
    }, [])

    const setFilterProjects = e => {
        setFilter(e.target.value);
    }

    const filterProjects = project => {
        return (filter === 'All') ? true : project.filters.includes(filter);
    }
    
    const projects = [
        {
            title: 'Portfolio',
            date: 'December 2016 - January 2022',
            src: '/img/portfolio-projects-v4b.png',
            alt: 'Portfolio',
            link: '/projects/portfolio',
            filters: ['JS', 'React', 'Web'],
            description: 'Personal portfolio website created to showcase my projects and technical skills. First created in 2016 as a simple static HTML website, the website has experienced multiple improvements over the years including migration to PHP, then to React with Materialize, and most recently to React with Tailwind.'
        },
        {
            title: 'LucidLab',
            date: 'October 2019 - May 2020',
            src: '/img/lucidlab1.jpg',
            alt: 'LucidLab',
            link: '/projects/lucidlab',
            filters: ['JS', 'Python', 'React', 'SQL', 'Group', 'Web'],
            description: 'An adaptable, heterogeneous IoT testbed for which I developed the front-end user interface. The UI allows registered users to upload and configure tests and images to be deployed on the testbed. Additionally, test results and custom metrics including mote availability graphs and mote CCA charts can be viewed.'
        },
        {
            title: 'Logical Sudoku Solver',
            date: 'November 2019 - January 2020',
            src: '/img/sudoku1.jpg',
            alt: 'Logical sudoku solver',
            link: '/projects/sudoku',
            filters: ['Python', 'Games'],
            description: 'Program that can solve expert level Sudoku using only logical techniques (no brute forcing, guessing or backtracking). The program outputs a detailed description of the techniques and moves required at each step to solve Sudoku. The solver reads CSV files allowing for batch solving. In-depth analysis is displayed including but not limited to: difficulty rating, occurrences/probability of each technique, processing time for each technique and total processing time.'
        },
        {
            title: 'React Minesweeper',
            date: 'October 2019',
            src: '/img/react-minesweeper1.jpg',
            alt: 'React Minesweeper',
            link: '/projects/react-minesweeper',
            filters: ['JS', 'React', 'Games', 'Web'],
            description: 'Clone of classic Minesweeper with an 18x18 board. Ability to reveal tiles and flag tiles for potential mines. Flood fill algorithms will reveal all adjacent tiles when a 0 is uncovered, mimicing the behaviour of the original game.'
        },
        {
            title: 'QuRVe',
            date: 'June 2019 - August 2019',
            src: '/img/qurve1.jpg',
            alt: 'QuRVe',
            link: '/projects/qurve',
            filters: ['JS', 'React', 'Web'],
            description: 'Designed and created an updated UI for the application QuRVe, a credit tech tool using JavaScript, Webix, React and Less. Implemented an interface to retrieve financial instruments data with a configurable dashboard showing graphs, pivot tables and descriptions of selected securities.'
        },
        {
            title: 'Using Swarm AI to Map a Cave Network',
            date: 'December 2018 - April 2019',
            src: '/img/cave-exploration1.jpg',
            alt: 'Cave Exploration',
            link: '/projects/cave-exploration',
            filters: ['C++'],
            description: 'Cave exploration is dangerous and time-consuming. This project demonstrates how swarm AI could be used in a group of autonomous flying drones to navigate and explore a cave as efficiently as possible. The project also includes random, realistic cave environment generation.'
        },
        {
            title: 'Roller Coaster',
            date: 'December 2018 - January 2019',
            src: '/img/roller-coaster1.jpg',
            alt: 'Roller coaster',
            link: '/projects/roller-coaster',
            filters: ['C++'],
            description: 'Simulation of a 3-car roller coaster which traverses a small looped track featuring a lift hill, drop, loop-the-loop and turns. The coaster can be viewed from a first-person perspective in each of the carts, as well as additional views that track the carts round the track.'
        },
        {
            title: 'RSCBot',
            date: 'February 2018 - March 2019',
            src: '/img/rscbot1.jpg',
            alt: 'RSCBot',
            link: '/projects/rscbot',
            filters: ['JS', 'PHP', 'SQL', 'Group', 'Web'],
            description: 'A specialised, personalisable trader chatbot that can fetch data and current news on financial stocks and sectors from the FTSE 100 index. RSCBot communicates with the user in a natural way, and has been designed to identify and adapt to the user\'s main interests in the stock market, making for a more personal trading experience.'
        },
        {
            title: 'URL Shortener',
            date: 'April 2017',
            src: '/img/url-shortener1.jpg',
            alt: 'URL shortener',
            link: '/projects/url-shortener',
            filters: ['JS', 'Web'],
            description: 'A URL shortening tool creating using Node.js, Express.js for routing GET and POST requests and MongoDB to store URLs. The project is my first using all three of these technologies. Long URLs are stored in the database and the ID is encoded and used as the short URL.'
        },
        {
            title: 'Graph Algorithm Visualiser',
            date: 'April 2017',
            src: '/img/graph-algorithm-visualiser1.jpg',
            alt: 'Graph algorithm visualiser',
            link: '/projects/graph-algorithm-visualiser',
            filters: ['JS', 'Web'],
            description: 'Web application which visualises the process of various graph algorithms on randomised nodes. The application can simulate Prim\'s and Kruskal\'s algorithm for constructing minimum spanning trees, Graham scan to construct convex hulls, Greedy algorithm to construct minimal matchings, and Nearest neighbour with 2-Opt to create Hamiltonian cycles.'
        },
        {
            title: 'Cavern Minesweeper',
            date: 'March 2017',
            src: '/img/cavern-minesweeper1.jpg',
            alt: 'Cavern Minesweeper',
            link: '/projects/cavern-minesweeper',
            filters: ['Games'],
            description: 'Cavern Minesweeper is based off of the original Minesweeper game but with tiered mines (ores) which can only be mined if the player has uncovered enough squares to advance to the required mining level. The game has three levels of difficulty to challenge competitive players. The project served as a tool to teach me Ruby.',
        },
        {
            title: 'BSplit',
            date: 'February 2017 - March 2017',
            src: '/img/bsplit1.jpg',
            alt: 'BSplit',
            link: '/projects/bsplit',
            filters: ['JS', 'PHP', 'SQL', 'Web'],
            description: 'BSplit is a web application that allows registered users to create and settle payments between friends and housemates for things such as restaurant bills, utility bills and food shopping. The app\'s dashboard utilises AJAX requests to prevent web page reloading to enchance user experience. Dashboard and email notifications are sent when a new bill or group is created.'
        },
        {
            title: 'To Do List',
            date: 'January 2017 - February 2017',
            src: '/img/todo-list1.jpg',
            alt: 'To Do List',
            link: '/projects/todo-list',
            filters: ['JS', 'PHP', 'SQL', 'Group', 'Web'],
            description: 'To Do List Web application where users can signup and create their own lists. Learned and integrated PHP and SQL to store and query user\'s lists. Protected the application from SQL injection and cross-site scripting.',
        },
        {
            title: 'Delivery Route Planner',
            date: 'December 2015 - March 2016',
            src: '/img/delivery-route-planner1.jpg',
            alt: 'Delivery route planner',
            link: '/projects/delivery-route-planner',
            filters: ['C#'],
            description: 'Software written in C# which creates an efficient route between multiple delivery locations. Uses SQL to store and query data on products, clients and their deliveries. Produces a report which communicates the route, all items in the deliveries, and the estimated time taken. Uses graph algorithms such as Dijkstra\'s and Nearest Neighbour with 2-Opt to optimise the route.',
        },
        {
            title: 'Sorting Algorithm Visualiser',
            date: 'September 2015 - November 2015',
            src: '/img/sorting-algorithm-visualiser1.jpg',
            alt: 'Sorting Algorithm Visualiser',
            link: '/projects/sorting-algorithm-visualiser',
            filters: ['C#'],
            description: 'Software tool for visualising the sorting process on generated datasets. Supports 20 different sorting algorithms including: Bubble sort, Merge sort, and Quicksort. The tool also has the ability to compare and visualise two sorting algorithms concurrently. The initial dataset can be a choice of 18 configurations such as: random, normally distributed or sawtooth.',
        }
    ];

    return (
        <>
        <Section>
            <h1 className='project-title mb-12 text-center md:text-left'>Projects</h1>
            
            <p className='text-sm mb-2 ml-2 font-bold text-link text-center md:text-left'>Filter Projects</p>
            <div className='flex md:block md:justify-start'>
                <select onChange={setFilterProjects} name='project-filter' className='mx-auto px-8 py-4 text-lg text-center cursor-pointer bg-white ring-2 ring-link focus:ring-link-hover rounded'>
                    <option value='All'>All</option>
                    <option value='C#'>C#</option>
                    <option value='C++'>C++</option>
                    <option value='JS'>JavaScript</option>
                    <option value='PHP'>PHP</option>
                    <option value='Python'>Python</option>
                    <option value='React'>React</option>
                    <option value='SQL'>SQL</option>
                    <option value='Games'>Games/Puzzles</option>
                    <option value='Group'>Group Projects</option>
                    <option value='Web'>Web Applications</option>
                </select>
            </div>
        </Section>

        {projects.filter(filterProjects).map((project, i) => {
            return (
                <ProjectItem
                    key={i}
                    title={project.title}
                    date={project.date}
                    description={project.description}
                    src={project.src}
                    alt={project.alt}
                    link={project.link}
                />
            )
        })}
        </>
    );
}
