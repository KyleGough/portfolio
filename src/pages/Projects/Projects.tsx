import React, { useEffect, useState } from 'react';

import { Divider } from '../../components/Divider';
import { ProjectFilter } from '../../components/ProjectFilter';
import { ProjectItem } from '../../components/ProjectItem';
import { Section } from '../../components/Section';
import { getAllProjects,IProject } from '../../data';

export const Projects: React.FC = () => {
  const [filter, setFilter] = useState('All');

  useEffect(() => {
    document.title = 'Project List - Kyle Gough';
  }, []);

  const filterProjects = (project: IProject): boolean => {
    return filter === 'All' ? true : project.filters.includes(filter);
  };

  return (
    <>
      <Section>
        <h1 className="project-title mb-8 text-center md:text-left">
          Projects
        </h1>
      </Section>

      <hr className="mt-8 h-px bg-divider" />

      <div className="bg-circuits">
        <Section overrideTopPadding={true}>
          <ProjectFilter filter={filter} setFilterCallback={setFilter} />
        </Section>
        <hr className="mb-8 h-px bg-divider" />
      </div>

      {Object.values(getAllProjects())
        .filter((project) => filterProjects(project))
        .map((project, i) => {
          return (
            <React.Fragment key={project.title}>
              {i !== 0 && <Divider />}
              <ProjectItem
                title={project.title}
                date={project.date}
                description={project.description}
                src={project.src}
                alt={project.alt}
                link={project.link}
              />
            </React.Fragment>
          );
        })}
    </>
  );
};
