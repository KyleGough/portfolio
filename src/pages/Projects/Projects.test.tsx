import { render } from '@testing-library/react';
import React from 'react';
import { BrowserRouter } from 'react-router-dom';

import { Projects } from './Projects';

describe('Projects component', () => {
  it('renders', async () => {
    const { container } = render(
      <BrowserRouter>
        <Projects />
      </BrowserRouter>
    );
    expect(container).toMatchSnapshot();
  });
});
