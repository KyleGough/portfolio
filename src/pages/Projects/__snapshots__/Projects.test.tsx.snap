// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Projects component renders 1`] = `
<div>
  <div
    class="first:pt-24 container text-primary py-8 last:pb-16"
  >
    <h1
      class="project-title mb-8 text-center md:text-left"
    >
      Projects
    </h1>
  </div>
  <hr
    class="mt-8 h-px bg-divider"
  />
  <div
    class="bg-circuits"
  >
    <div
      class="container text-primary py-8 last:pb-16"
    >
      <p
        class="text-sm mb-2 ml-2 font-bold text-link text-center md:text-left"
      >
        Filter Projects
      </p>
      <div
        class="relative w-64 mx-auto md:mx-0"
      >
        <button
          aria-controls="project-filter-selection"
          aria-expanded="false"
          aria-haspopup="listbox"
          class="group relative w-full pl-6 pr-12 py-4 cursor-pointer bg-background ring-0 rounded-sm shadow border-light"
        >
          <span
            class="text-link group-hover:text-link-hover float-left"
          >
            All
          </span>
          <svg
            class="w-8 h-8 absolute top-[12px] right-0 flex items-center pr-2 pointer-events-none fill-link group-hover:fill-link-hover"
            height="24px"
            viewBox="0 0 24 24"
            width="24px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M12 5.83L15.17 9l1.41-1.41L12 3 7.41 7.59 8.83 9 12 5.83zm0 12.34L8.83 15l-1.41 1.41L12 21l4.59-4.59L15.17 15 12 18.17z"
            />
          </svg>
        </button>
      </div>
    </div>
    <hr
      class="mb-8 h-px bg-divider"
    />
  </div>
  <div
    class="first:pt-24 container text-primary py-8 last:pb-16"
  >
    <div
      class="grid grid-cols-12 gap-8 lg:gap-16"
    >
      <div
        class="col-span-12 lg:col-span-7"
      >
        <h2
          class="mb-4 project-header"
        >
          Portfolio
        </h2>
        <p
          class="my-4 text-link-hover"
        >
          December 2016 - September 2022
        </p>
        <p
          class="my-4"
        >
          Personal portfolio website created to showcase my projects and technical skills. First created in 2016 as a simple static HTML website, the website has experienced multiple improvements over the years including migration to PHP, then to React with Materialize, and most recently to React with Tailwind.
        </p>
      </div>
      <div
        class="col-span-12 lg:col-span-5 text-center -mb-8"
      >
        <a
          class="group relative"
          href="/projects/portfolio"
        >
          <img
            alt="Portfolio"
            class="border-2 border-white rounded-lg shadow"
            loading="lazy"
            src="/img/card-portfolio.jpg"
          />
          <button
            class="relative -top-[1.6rem] px-12 py-3 bg-background text-link group-hover:text-link-hover group-focus:text-link-hover rounded-sm shadow border-light whitespace-nowrap"
            tabindex="-1"
          >
            View Project
          </button>
        </a>
      </div>
    </div>
  </div>
  <hr
    class="m-8 h-px bg-divider"
  />
  <div
    class="first:pt-24 container text-primary py-8 last:pb-16"
  >
    <div
      class="grid grid-cols-12 gap-8 lg:gap-16"
    >
      <div
        class="col-span-12 lg:col-span-7"
      >
        <h2
          class="mb-4 project-header"
        >
          LucidLab
        </h2>
        <p
          class="my-4 text-link-hover"
        >
          October 2019 - May 2020
        </p>
        <p
          class="my-4"
        >
          An adaptable, heterogeneous IoT testbed for which I developed the front-end user interface. The UI allows registered users to upload and configure tests and images to be deployed on the testbed. Additionally, test results and custom metrics including mote availability graphs and mote CCA charts can be viewed.
        </p>
      </div>
      <div
        class="col-span-12 lg:col-span-5 text-center -mb-8"
      >
        <a
          class="group relative"
          href="/projects/lucidlab"
        >
          <img
            alt="LucidLab"
            class="border-2 border-white rounded-lg shadow"
            loading="lazy"
            src="/img/card-lucidlab.jpg"
          />
          <button
            class="relative -top-[1.6rem] px-12 py-3 bg-background text-link group-hover:text-link-hover group-focus:text-link-hover rounded-sm shadow border-light whitespace-nowrap"
            tabindex="-1"
          >
            View Project
          </button>
        </a>
      </div>
    </div>
  </div>
  <hr
    class="m-8 h-px bg-divider"
  />
  <div
    class="first:pt-24 container text-primary py-8 last:pb-16"
  >
    <div
      class="grid grid-cols-12 gap-8 lg:gap-16"
    >
      <div
        class="col-span-12 lg:col-span-7"
      >
        <h2
          class="mb-4 project-header"
        >
          Logical Sudoku Solver
        </h2>
        <p
          class="my-4 text-link-hover"
        >
          November 2019 - January 2020
        </p>
        <p
          class="my-4"
        >
          Program that can solve expert level Sudoku using only logical techniques (no brute forcing, guessing or backtracking). The program outputs a detailed description of the techniques and moves required at each step to solve Sudoku. The solver reads CSV files allowing for batch solving. In-depth analysis is displayed including but not limited to: difficulty rating, occurrences of each technique, probability of each technique, processing time for each technique and total processing time.
        </p>
      </div>
      <div
        class="col-span-12 lg:col-span-5 text-center -mb-8"
      >
        <a
          class="group relative"
          href="/projects/sudoku"
        >
          <img
            alt="Logical sudoku solver"
            class="border-2 border-white rounded-lg shadow"
            loading="lazy"
            src="/img/card-sudoku.jpg"
          />
          <button
            class="relative -top-[1.6rem] px-12 py-3 bg-background text-link group-hover:text-link-hover group-focus:text-link-hover rounded-sm shadow border-light whitespace-nowrap"
            tabindex="-1"
          >
            View Project
          </button>
        </a>
      </div>
    </div>
  </div>
  <hr
    class="m-8 h-px bg-divider"
  />
  <div
    class="first:pt-24 container text-primary py-8 last:pb-16"
  >
    <div
      class="grid grid-cols-12 gap-8 lg:gap-16"
    >
      <div
        class="col-span-12 lg:col-span-7"
      >
        <h2
          class="mb-4 project-header"
        >
          React Minesweeper
        </h2>
        <p
          class="my-4 text-link-hover"
        >
          October 2019
        </p>
        <p
          class="my-4"
        >
          Clone of classic Minesweeper with an 18x18 board. Ability to reveal tiles and flag tiles for potential mines. Flood fill algorithms will reveal all adjacent tiles when a 0 is uncovered, mimicing the behaviour of the original game.
        </p>
      </div>
      <div
        class="col-span-12 lg:col-span-5 text-center -mb-8"
      >
        <a
          class="group relative"
          href="/projects/react-minesweeper"
        >
          <img
            alt="React Minesweeper"
            class="border-2 border-white rounded-lg shadow"
            loading="lazy"
            src="img/card-react-minesweeper.jpg"
          />
          <button
            class="relative -top-[1.6rem] px-12 py-3 bg-background text-link group-hover:text-link-hover group-focus:text-link-hover rounded-sm shadow border-light whitespace-nowrap"
            tabindex="-1"
          >
            View Project
          </button>
        </a>
      </div>
    </div>
  </div>
  <hr
    class="m-8 h-px bg-divider"
  />
  <div
    class="first:pt-24 container text-primary py-8 last:pb-16"
  >
    <div
      class="grid grid-cols-12 gap-8 lg:gap-16"
    >
      <div
        class="col-span-12 lg:col-span-7"
      >
        <h2
          class="mb-4 project-header"
        >
          Using Swarm AI to Map a Cave Network
        </h2>
        <p
          class="my-4 text-link-hover"
        >
          December 2018 - April 2019
        </p>
        <p
          class="my-4"
        >
          Cave exploration is dangerous and time-consuming. This project demonstrates how swarm AI could be used in a group of autonomous flying drones to navigate and explore a cave as efficiently as possible. The project also includes random, realistic cave environment generation.
        </p>
      </div>
      <div
        class="col-span-12 lg:col-span-5 text-center -mb-8"
      >
        <a
          class="group relative"
          href="/projects/cave-exploration"
        >
          <img
            alt="Cave Exploration"
            class="border-2 border-white rounded-lg shadow"
            loading="lazy"
            src="/img/card-cave-exploration.jpg"
          />
          <button
            class="relative -top-[1.6rem] px-12 py-3 bg-background text-link group-hover:text-link-hover group-focus:text-link-hover rounded-sm shadow border-light whitespace-nowrap"
            tabindex="-1"
          >
            View Project
          </button>
        </a>
      </div>
    </div>
  </div>
  <hr
    class="m-8 h-px bg-divider"
  />
  <div
    class="first:pt-24 container text-primary py-8 last:pb-16"
  >
    <div
      class="grid grid-cols-12 gap-8 lg:gap-16"
    >
      <div
        class="col-span-12 lg:col-span-7"
      >
        <h2
          class="mb-4 project-header"
        >
          Roller Coaster
        </h2>
        <p
          class="my-4 text-link-hover"
        >
          December 2018 - January 2019
        </p>
        <p
          class="my-4"
        >
          Simulation of a 3-car roller coaster which traverses a small looped track featuring a lift hill, drop, loop-the-loop and turns. The coaster can be viewed from a first-person perspective in each of the carts, as well as additional views that track the carts round the track.
        </p>
      </div>
      <div
        class="col-span-12 lg:col-span-5 text-center -mb-8"
      >
        <a
          class="group relative"
          href="/projects/roller-coaster"
        >
          <img
            alt="Roller coaster"
            class="border-2 border-white rounded-lg shadow"
            loading="lazy"
            src="/img/card-roller-coaster.jpg"
          />
          <button
            class="relative -top-[1.6rem] px-12 py-3 bg-background text-link group-hover:text-link-hover group-focus:text-link-hover rounded-sm shadow border-light whitespace-nowrap"
            tabindex="-1"
          >
            View Project
          </button>
        </a>
      </div>
    </div>
  </div>
  <hr
    class="m-8 h-px bg-divider"
  />
  <div
    class="first:pt-24 container text-primary py-8 last:pb-16"
  >
    <div
      class="grid grid-cols-12 gap-8 lg:gap-16"
    >
      <div
        class="col-span-12 lg:col-span-7"
      >
        <h2
          class="mb-4 project-header"
        >
          RSCBot
        </h2>
        <p
          class="my-4 text-link-hover"
        >
          February 2018 - March 2019
        </p>
        <p
          class="my-4"
        >
          A specialised, personalisable trader chatbot that can fetch data and current news on financial stocks and sectors from the FTSE 100 index. RSCBot communicates with the user in a natural way, and has been designed to identify and adapt to the user's main interests in the stock market, making for a more personal trading experience.
        </p>
      </div>
      <div
        class="col-span-12 lg:col-span-5 text-center -mb-8"
      >
        <a
          class="group relative"
          href="/projects/rscbot"
        >
          <img
            alt="RSCBot"
            class="border-2 border-white rounded-lg shadow"
            loading="lazy"
            src="/img/card-rscbot.jpg"
          />
          <button
            class="relative -top-[1.6rem] px-12 py-3 bg-background text-link group-hover:text-link-hover group-focus:text-link-hover rounded-sm shadow border-light whitespace-nowrap"
            tabindex="-1"
          >
            View Project
          </button>
        </a>
      </div>
    </div>
  </div>
  <hr
    class="m-8 h-px bg-divider"
  />
  <div
    class="first:pt-24 container text-primary py-8 last:pb-16"
  >
    <div
      class="grid grid-cols-12 gap-8 lg:gap-16"
    >
      <div
        class="col-span-12 lg:col-span-7"
      >
        <h2
          class="mb-4 project-header"
        >
          URL Shortener
        </h2>
        <p
          class="my-4 text-link-hover"
        >
          April 2017
        </p>
        <p
          class="my-4"
        >
          A URL shortening tool creating using Node.js, Express.js for routing GET and POST requests and MongoDB to store URLs. The project is my first using all three of these technologies. Long URLs are stored in the database and the ID is encoded and used as the short URL.
        </p>
      </div>
      <div
        class="col-span-12 lg:col-span-5 text-center -mb-8"
      >
        <a
          class="group relative"
          href="/projects/url-shortener"
        >
          <img
            alt="URL shortener"
            class="border-2 border-white rounded-lg shadow"
            loading="lazy"
            src="/img/card-url-shortener.jpg"
          />
          <button
            class="relative -top-[1.6rem] px-12 py-3 bg-background text-link group-hover:text-link-hover group-focus:text-link-hover rounded-sm shadow border-light whitespace-nowrap"
            tabindex="-1"
          >
            View Project
          </button>
        </a>
      </div>
    </div>
  </div>
  <hr
    class="m-8 h-px bg-divider"
  />
  <div
    class="first:pt-24 container text-primary py-8 last:pb-16"
  >
    <div
      class="grid grid-cols-12 gap-8 lg:gap-16"
    >
      <div
        class="col-span-12 lg:col-span-7"
      >
        <h2
          class="mb-4 project-header"
        >
          Graph Algorithm Visualiser
        </h2>
        <p
          class="my-4 text-link-hover"
        >
          April 2017
        </p>
        <p
          class="my-4"
        >
          Web application which visualises the process of various graph algorithms on randomised nodes. The application can simulate Prim's and Kruskal's algorithm for constructing minimum spanning trees, Graham scan to construct convex hulls, Greedy algorithm to construct minimal matchings, and Nearest neighbour with 2-Opt to create Hamiltonian cycles.
        </p>
      </div>
      <div
        class="col-span-12 lg:col-span-5 text-center -mb-8"
      >
        <a
          class="group relative"
          href="/projects/graph-algorithm-visualiser"
        >
          <img
            alt="Graph algorithm visualiser"
            class="border-2 border-white rounded-lg shadow"
            loading="lazy"
            src="/img/card-graph-algorithm-visualiser.jpg"
          />
          <button
            class="relative -top-[1.6rem] px-12 py-3 bg-background text-link group-hover:text-link-hover group-focus:text-link-hover rounded-sm shadow border-light whitespace-nowrap"
            tabindex="-1"
          >
            View Project
          </button>
        </a>
      </div>
    </div>
  </div>
  <hr
    class="m-8 h-px bg-divider"
  />
  <div
    class="first:pt-24 container text-primary py-8 last:pb-16"
  >
    <div
      class="grid grid-cols-12 gap-8 lg:gap-16"
    >
      <div
        class="col-span-12 lg:col-span-7"
      >
        <h2
          class="mb-4 project-header"
        >
          Cavern Minesweeper
        </h2>
        <p
          class="my-4 text-link-hover"
        >
          March 2017
        </p>
        <p
          class="my-4"
        >
          Cavern Minesweeper is based off of the original Minesweeper game but with tiered mines (ores) which can only be mined if the player has uncovered enough squares to advance to the required mining level. The game has three levels of difficulty to challenge competitive players. The project served as a tool to teach me Ruby.
        </p>
      </div>
      <div
        class="col-span-12 lg:col-span-5 text-center -mb-8"
      >
        <a
          class="group relative"
          href="/projects/cavern-minesweeper"
        >
          <img
            alt="Cavern Minesweeper"
            class="border-2 border-white rounded-lg shadow"
            loading="lazy"
            src="/img/card-cavern-minesweeper.jpg"
          />
          <button
            class="relative -top-[1.6rem] px-12 py-3 bg-background text-link group-hover:text-link-hover group-focus:text-link-hover rounded-sm shadow border-light whitespace-nowrap"
            tabindex="-1"
          >
            View Project
          </button>
        </a>
      </div>
    </div>
  </div>
  <hr
    class="m-8 h-px bg-divider"
  />
  <div
    class="first:pt-24 container text-primary py-8 last:pb-16"
  >
    <div
      class="grid grid-cols-12 gap-8 lg:gap-16"
    >
      <div
        class="col-span-12 lg:col-span-7"
      >
        <h2
          class="mb-4 project-header"
        >
          BSplit
        </h2>
        <p
          class="my-4 text-link-hover"
        >
          February 2017 - March 2017
        </p>
        <p
          class="my-4"
        >
          BSplit is a web application that allows registered users to create and settle payments between friends and housemates for things such as restaurant bills, utility bills and food shopping. The app's dashboard utilises AJAX requests to prevent web page reloading to enchance user experience. Dashboard and email notifications are sent when a new bill or group is created.
        </p>
      </div>
      <div
        class="col-span-12 lg:col-span-5 text-center -mb-8"
      >
        <a
          class="group relative"
          href="/projects/bsplit"
        >
          <img
            alt="BSplit"
            class="border-2 border-white rounded-lg shadow"
            loading="lazy"
            src="/img/card-bsplit.jpg"
          />
          <button
            class="relative -top-[1.6rem] px-12 py-3 bg-background text-link group-hover:text-link-hover group-focus:text-link-hover rounded-sm shadow border-light whitespace-nowrap"
            tabindex="-1"
          >
            View Project
          </button>
        </a>
      </div>
    </div>
  </div>
  <hr
    class="m-8 h-px bg-divider"
  />
  <div
    class="first:pt-24 container text-primary py-8 last:pb-16"
  >
    <div
      class="grid grid-cols-12 gap-8 lg:gap-16"
    >
      <div
        class="col-span-12 lg:col-span-7"
      >
        <h2
          class="mb-4 project-header"
        >
          To Do List
        </h2>
        <p
          class="my-4 text-link-hover"
        >
          January 2017 - February 2017
        </p>
        <p
          class="my-4"
        >
          To Do List Web application where users can signup and create their own lists. Learned and integrated PHP and SQL to store and query user's lists. Protected the application from SQL injection and cross-site scripting.
        </p>
      </div>
      <div
        class="col-span-12 lg:col-span-5 text-center -mb-8"
      >
        <a
          class="group relative"
          href="/projects/todo-list"
        >
          <img
            alt="To Do List"
            class="border-2 border-white rounded-lg shadow"
            loading="lazy"
            src="/img/card-todo.jpg"
          />
          <button
            class="relative -top-[1.6rem] px-12 py-3 bg-background text-link group-hover:text-link-hover group-focus:text-link-hover rounded-sm shadow border-light whitespace-nowrap"
            tabindex="-1"
          >
            View Project
          </button>
        </a>
      </div>
    </div>
  </div>
  <hr
    class="m-8 h-px bg-divider"
  />
  <div
    class="first:pt-24 container text-primary py-8 last:pb-16"
  >
    <div
      class="grid grid-cols-12 gap-8 lg:gap-16"
    >
      <div
        class="col-span-12 lg:col-span-7"
      >
        <h2
          class="mb-4 project-header"
        >
          Delivery Route Planner
        </h2>
        <p
          class="my-4 text-link-hover"
        >
          December 2015 - March 2016
        </p>
        <p
          class="my-4"
        >
          Creates an efficient route between multiple delivery locations. Store, view and query data on products, clients and their deliveries. Produces a report which communicates the route, delivery items, and ETA. Uses graph algorithms such as Dijkstra's and Nearest Neighbour with 2-Opt to optimise the route.
        </p>
      </div>
      <div
        class="col-span-12 lg:col-span-5 text-center -mb-8"
      >
        <a
          class="group relative"
          href="/projects/delivery-route-planner"
        >
          <img
            alt="Delivery route planner"
            class="border-2 border-white rounded-lg shadow"
            loading="lazy"
            src="/img/card-delivery-route-planner.jpg"
          />
          <button
            class="relative -top-[1.6rem] px-12 py-3 bg-background text-link group-hover:text-link-hover group-focus:text-link-hover rounded-sm shadow border-light whitespace-nowrap"
            tabindex="-1"
          >
            View Project
          </button>
        </a>
      </div>
    </div>
  </div>
  <hr
    class="m-8 h-px bg-divider"
  />
  <div
    class="first:pt-24 container text-primary py-8 last:pb-16"
  >
    <div
      class="grid grid-cols-12 gap-8 lg:gap-16"
    >
      <div
        class="col-span-12 lg:col-span-7"
      >
        <h2
          class="mb-4 project-header"
        >
          Sorting Algorithm Visualiser
        </h2>
        <p
          class="my-4 text-link-hover"
        >
          September 2015 - November 2015
        </p>
        <p
          class="my-4"
        >
          Tool for visualising the sorting process on generated datasets. Supports 20 different sorting algorithms including: Bubble sort, Merge sort, and Quicksort. The tool also has the ability to compare and visualise two sorting algorithms concurrently. The initial dataset can be a choice of 18 configurations such as: random, normally distributed or sawtooth.
        </p>
      </div>
      <div
        class="col-span-12 lg:col-span-5 text-center -mb-8"
      >
        <a
          class="group relative"
          href="/projects/sorting-algorithm-visualiser"
        >
          <img
            alt="Sorting Algorithm Visualiser"
            class="border-2 border-white rounded-lg shadow"
            loading="lazy"
            src="/img/card-sorting-algorithm-visualiser.jpg"
          />
          <button
            class="relative -top-[1.6rem] px-12 py-3 bg-background text-link group-hover:text-link-hover group-focus:text-link-hover rounded-sm shadow border-light whitespace-nowrap"
            tabindex="-1"
          >
            View Project
          </button>
        </a>
      </div>
    </div>
  </div>
</div>
`;
